### ISSUE TITLE
ValueError: signal only works in main thread of the main interpreter

### ISSUE THREAD

**Issue by mriffle** [Slimmer-AI/mpire#44]
I'm running into the following error running on Python 3.10.4 in the official Python docker image pulled down from Docker Hub.

```
Traceback (most recent call last):
  File "/usr/src/app/app/request_handler.py", line 88, in process_request
    for result_dict in pool.imap_unordered(
  File "/usr/local/lib/python3.10/site-packages/mpire/pool.py", line 542, in imap_unordered
    self._start_workers(progress_bar)
  File "/usr/local/lib/python3.10/site-packages/mpire/pool.py", line 148, in _start_workers
    self._workers.append(self._start_worker(worker_id))
  File "/usr/local/lib/python3.10/site-packages/mpire/pool.py", line 208, in _start_worker
    with DisableKeyboardInterruptSignal():
  File "/usr/local/lib/python3.10/site-packages/mpire/signal.py", line 37, in __enter__
    ignore_keyboard_interrupt()
  File "/usr/local/lib/python3.10/site-packages/mpire/signal.py", line 45, in ignore_keyboard_interrupt
    signal_(SIGINT, SIG_IGN)
  File "/usr/local/lib/python3.10/signal.py", line 56, in signal
    handler = _signal.signal(_enum_to_int(signalnum), _enum_to_int(handler))
ValueError: signal only works in main thread of the main interpreter
```

Googling this error wasn't helpful. Is this related to running in Docker, or more likely an error I'm making?

Here is the bit of code that runs imap_unordered():

```
        with WorkerPool(n_jobs=4, pass_worker_id=False) as pool:
            for result_dict in pool.imap_unordered(
                    create_ms2_file,
                    zip(request_data, range(1, len(request_data) + 1), [workdir] * len(request_data)),
                    iterable_len=len(request_data),
                    progress_bar=False
            ):
                percent_done += percent_per_file
                request_status_dict[request['id']]['end_user_message'] = 'Exporting scan files: ' +\
                                                                         str(round(percent_done, 1)) +\
                                                                         '% complete...'

                result_dicts[result_dict['spectr_file_id']] = result_dict
```

It _looks_ to me like that arguments being passed to pool.imap_unordered are stucutured correctly.

Any guidance is appreciated, thanks.

**Comment by sybrenjansen** [comment]
You're arguments look fine indeed, although I think you could benefit from using the shared_objects functionality to speed things up (have a look in the docs).

Getting back to your issue, given the first part of your traceback it looks like you're running this code in some kind of request handler, which probably uses another tread than the main thread. Can you verify this? 

In mpire, when workers are started, the keyboard interrupt signal is disabled such that it can do a graceful shutdown in case this signal is issued. Doing stuff with signals only work in the main thread and I assumed that workers are always started there. In your example, this assumption isn't true it seems, so I will need to revise the code a bit to only disable the signal in case we're living in the main thread. 

I.e. You found yourself a bug. It should be quite an easy bug to fix though. I will see if I can find some time early next week

**Comment by mriffle** [comment]
Thanks for the quick response! Indeed, I do use Python's `threading` library to call a function in a new thread that subsequently calls this code. 

Thanks for the tip on the shared_objects functionality, will check that out.

**Comment by sybrenjansen** [comment]
Should be fixed in v2.4.0 which is building right now and should be available in a minute or two

**Comment by mriffle** [comment]
Looks to be working great now, thank you
